src/target/mspm0.c

```diff
@@ -0,0 +1,349 @@
+	/* {"write_nm", mspm0_write_non_main, "Write data to NONMAIN (configuration) flash. write_nm <address> <value>"}, */
+	{NULL, NULL, NULL}};
+
+typedef struct conf_register {
+	uint16_t offset : 12;
+	uint16_t words : 4;
+	char const *id;
+} conf_register_s;
+
+static conf_register_s mspm0_factory_regs[] = {{0x00, 1, "TRACEID"}, {0x04, 1, "DEVICEID"}, {0x08, 1, "USERID"},
+	{0x0c, 1, "BSLPIN_UART"}, {0x10, 1, "BSLPIN_I2C"}, {0x14, 1, "BSLPIN_INVOKE"}, {0x18, 1, "SRAMFLASH"},
+	{0x3c, 1, "TEMP_SENSE0"}, {0x7c, 1, "BOOTCRC"}, {0, 0, NULL}};
+
+static conf_register_s mspm0_bcr_regs[] = {{0x00, 1, "BCRCONFIGID"}, {0x04, 1, "BOOTCFG0"}, {0x08, 1, "BOOTCFG1"},
+	{0x0c, 4, "PWDDEBUGLOCK"}, {0x1c, 4, "BOOTCFG2"}, {0x20, 1, "BOOTCFG3"}, {0x24, 4, "PWDMASSERASE"},
+	{0x34, 4, "PWDFACTORYRESET"}, {0x44, 1, "FLASHSWP0"}, {0x48, 1, "FLASHSWP1"}, {0x4c, 1, "BOOTCFG4"},
+	{0x50, 1, "APPCRCSTART"}, {0x54, 1, "APPCRCLENGTH"}, {0x58, 1, "APPCRC"}, {0x5c, 1, "BOOTCRC"},
+	{0x100, 1, "BSLCONFIGID"}, {0x104, 1, "BSLPINCFG0"}, {0x108, 1, "BSLPINCFG1"}, {0x10c, 1, "BSLCONFIG0"},
+	{0x110, 8, "BSLPW"}, {0x130, 1, "BSLPLUGINCFG"}, {0x134, 4, "BSLPLUGINHOOK"}, {0x144, 1, "PATCHHOOKID"},
+	{0x148, 1, "SBLADDRESS"}, {0x14c, 1, "BSLAPPVER"}, {0x150, 1, "BSLCONFIG1"}, {0x154, 1, "BSLCRC"}, {0, 0, NULL}};
+
+static void mspm0_dump_regs(target_s *const target, const conf_register_s *regs, uint32_t base)
+{
+	for (conf_register_s const *r = regs; r->id; ++r) {
+		tc_printf(target, "%15s: ", r->id);
+		for (int q = 0; q < r->words; ++q) {
+			uint32_t value = target_mem32_read32(target, base + r->offset + q * 4);
+			tc_printf(target, "0x%08" PRIx32 "%s", value, q == r->words - 1 ? "\n" : " ");
+		}
+	}
+}
+
+static bool mspm0_dump_factory_config(target_s *const target, const int argc, const char **const argv)
+{
+	(void)argc;
+	(void)argv;
+	mspm0_dump_regs(target, mspm0_factory_regs, MSPM0_FLASH_FACTORY);
+	return true;
+}
+
+static bool mspm0_dump_bcr_config(target_s *const target, const int argc, const char **const argv)
+{
+	(void)argc;
+	(void)argv;
+	mspm0_dump_regs(target, mspm0_bcr_regs, MSPM0_FLASH_NONMAIN);
+	return true;
+}
+#endif
+
+static void mspm0_add_flash(target_s *const target, const uint32_t base, const size_t length, uint32_t banks)
+{
+	mspm0_flash_s *const flash = calloc(1, sizeof(*flash));
+	if (flash == NULL) {
+		DEBUG_WARN("%s: calloc failed for %" PRIu32 " bytes\n", __func__, (uint32_t)length);
+		return;
+	}
+
+	flash->banks = banks;
+	target_flash_s *target_flash = &flash->target_flash;
+	target_flash->start = base;
+	target_flash->length = length;
+	target_flash->blocksize = MSPM0_FLASH_SECTOR_SZ;
+	target_flash->writesize = 8U;
+	target_flash->erase = mspm0_flash_erase;
+	target_flash->write = mspm0_flash_write;
+	target_flash->erased = 0xff;
+	target_add_flash(target, target_flash);
+}
+
+#define BIT_LMASK(T, bits) ((T)-1) >> (sizeof(T) * CHAR_BIT - (bits))
+#define BITS(h, l, x)      (uint32_t)((x >> l) & BIT_LMASK(uint32_t, h - l + 1))
```

Comment:
This tries too be clever, but it falls foul of two issues which are going to cause eminent pain to maintain here: CHAR_BIT is platform-dependant, so all the shifts this tries to calculate are also platform-dependant; This performs more signed-unsigned conversions due to unsuffixed constants (1 is signed, 1U is unsigned), and may well invoke undefined behaviour (though we can't be certain) with the shifts being defined. This is liable to entirely miscompiling when building BMDA on spicier architectures.

Please instead do as the rest of the code base does and define `MSPM0_` prefixed mask and shift macros and apply those below. This then defines what the bit ranges mean with a friendly name, and is not liable to explode at a moment's notice down the road.